### ISSUE TITLE
[Pytorch][ROCm 7.14] Qwen3 MoE TP+EP training produces extremely large(nan) gradient norms

### ISSUE THREAD

**Issue by vidushi8** [pytorch/pytorch#194004]
## Problem

The ROCm 7.14 PyTorch wheel produces extremely large gradient
norms during Qwen3 MoE training with FSDP+TP+EP on torchtitan. The equivalent ROCm 7.2 wheel
produces stable gradient norms.

**In some runs this can lead to NaN values or loss-parity failures.**
**UPDATE: This issue has been fixed in upcoming rocm7.15/rocm10.0 release. Once a compatible pytorch whl is available; the fixed will be fetched to torchtitan repo.** 

## Failing test

ROCm 8-GPU feature test:

**Qwen3 MoE FSDP+TP+EP loss parity against reference**

Configuration:

- Model: `qwen3`
- Config: `qwen3_moe_debug`
- FSDP degree: 4
- Tensor parallel degree: 2
- Expert parallel degree: 4
- GPUs: 8
- Deterministic mode with seed 42
- GPU: AMD Instinct MI355X (`gfx950`)

## Reproduction command

```bash
MODULE=qwen3 CONFIG=qwen3_moe_debug NGPU=8 LOG_RANK=0 \
./run_train.sh \
  --parallelism.tensor_parallel_degree 2 \
  --parallelism.expert_parallel_degree 4 \
  --debug.deterministic \
  --debug.seed=42 \
  --metrics.enable_tensorboard \
  --metrics.log_freq=1 \
  --training.steps=100 \
  --metrics.save_tb_folder=tb_baseline \
  --checkpoint.enable \
  --checkpoint.export_dtype=bfloat16 \
  --checkpoint.load_only


ROCm 7.14 environment — failing
whl package: --index-url https://download.pytorch.org/whl/nightly/rocm7.14 "torch==2.14.0.dev20260805+rocm7.14"
torch:              2.14.0.dev20260805+rocm7.14
torch git:          f012e05e101851dce4c495559849318cbb7ed119
HIP:                7.14.60850
spmd_types:         0.2.1
triton-rocm:        3.8.0+git10f6be36
Example output:

step:  1  loss: 8.10514  grad_norm: 21485.1250
step:  5  loss: 5.08495  grad_norm: 66.5681
step:  6  loss: 4.94442  grad_norm: 62844.9609
step: 26  loss: 3.65015  grad_norm: 284010.5938
step: 41  loss: 3.43462  grad_norm: 49713.1133
step: 90  loss: 2.96572  grad_norm: 101188.0547

ROCm 7.2 environment — working
whl package: --index-url https://download.pytorch.org/whl/nightly/rocm7.2 "torch==2.14.0.dev20260802+rocm7.2"
torch:       2.14.0.dev20260802+rocm7.2
torch git:   944416f66ed4dd9502c3ff18a94881857ae6bab3
HIP:         7.2.53211
spmd_types:  0.2.1
triton-rocm: 3.8.0+git10f6be36
Example output:

step:  1  loss: 8.10514  grad_norm: 2.9300
step:  5  loss: 4.82658  grad_norm: 2.5541
step: 26  loss: 3.15735  grad_norm: 0.7451
step: 41  loss: 3.01872  grad_norm: 0.3288
step: 90  loss: 2.82196  grad_norm: 0.2069
Important observation
Both environments produce the same initial loss:

step 1 loss: 8.10514
However, the step-1 gradient norm differs significantly:

ROCm 7.2:   2.9300
ROCm 7.14:  21485.1250
The losses diverge after the first optimizer update:

ROCm 7.2 step 2 loss:   7.34490
ROCm 7.14 step 2 loss:  7.76990
This suggests the initialization, input data, and initial forward pass are equivalent, while the backward pass or distributed gradient-norm calculation differs.

**Expected behavior**
ROCm 7.14 should produce finite, stable gradient norms comparable to ROCm 7.2 and pass the Qwen3 MoE loss-parity test.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @jataylo @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang

**Comment by vidushi8** [comment]
# Summary: Qwen3 MoE TP=2, EP=4, FSDP Workload

## `.venv-0803` — ROCm 7.2

- **PyTorch:** Originally `2.14.0.dev20260803+rocm7.2` (commit `5b147731`); later upgraded to `2.14.0.dev20260805+rocm7.2` (commit `f012e05e`)
- **Result:** Both runs completed 100 steps with stable, identical gradients.

| Step | Loss | Grad norm |
|------|------|-----------|
| 1 | 8.10514 | 2.9300 |
| 41 | — | 0.3288 |
| 100 | 2.77590 | 0.2063 |

---

## `.venv-rocm7.14-whl` — ROCm 7.14

- **PyTorch:** `2.14.0.dev20260805+rocm7.14`
- **Commit:** `f012e05e101851dce4c495559849318cbb7ed119`
- **HIP:** 7.14.60850
- **ROCm:** 7.14.0
- **Result:** Produced deterministic extreme gradient spikes, including approximately **21485**, **62845**, **284011**, and **49713**.
- **`foreach=False` did not fix it.**

---

## `rocm/primus:v26.5` Docker — ROCm 7.15

- **Initial PyTorch:** `2.12.0+rocm7.15.0a20260720` — failed because `create_block_mask()` lacked `separate_full_blocks`
- **Upgraded PyTorch:** `2.14.0a0+rocm7.15.0a20260721` from [ROCm nightlies](https://rocm.nightlies.amd.com/whl-multi-arch/torch/)
- **Commit:** `3f8a4a186220e6772bd5f91027459808560d8539`
- **HIP:** 7.15.0
- **Result:** After removing the ABI-incompatible FBGEMM extension, the workload completed with stable gradients matching ROCm 7.2.

---

## Conclusion

The same PyTorch commit **`f012e05e`** is stable with **ROCm 7.2** but fails with **ROCm 7.14**. **ROCm 7.15** is also stable with a nearby PyTorch build.

This strongly localizes the regression to the **ROCm 7.14 wheel/runtime combination**, not the model, seed, `foreach` implementation, or PyTorch source commit alone.

---

# Triage Summary

## Ruled out

- **Attention backend is not responsible:** spikes occur with both FlexAttention and SDPA.
- **EP communication is not responsible:** spikes reproduce with EP=1.
- **`torch._grouped_mm` alone is not responsible:** BF16 reference per-expert matmuls also spike.
- **Dense Qwen3 remains stable,** confirming this is MoE-specific.
- **Uniform routing scores still spike,** ruling out router-score magnitude and its gradient path.

## Key findings

- **Balanced expert assignments eliminate spikes,** showing uneven dynamic expert workloads are required.
- **FP32 expert GEMMs eliminate spikes** while preserving dynamic routing.

## Conclusion

**ROCm 7.14’s BF16 expert GEMM path is numerically unstable for dynamically varying per-expert matrix shapes.**
